torchdata/datapipes/map/util/cacheholder.py

```diff
@@ -0,0 +1,46 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+from typing import Callable, Optional, Set, TypeVar
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.map import MapDataPipe
+
+
+T_co = TypeVar("T_co", covariant=True)
+
+
+@functional_datapipe("in_memory_cache")
+class InMemoryCacheHolderMapDataPipe(MapDataPipe[T_co]):
+    r"""
+    Stores elements from the source DataPipe in memory (functional name: ``in_memory_cache``).
+
+    The default ``cache`` is a ``Dict``. Depending on the use case and shape of the data, it may be useful to use other
+    objects as the cache, such as an LRU cache, ``numpy.array``, or ``multiprocessing.Array``.
+    to cache data in shared memory.
+
+    Args:
+        source_dp: source DataPipe from which elements are read and stored in memory
+        cache: a cache object that implements ``__getitem__`` and ``__setitem__``
+
+    Example:
+        >>> from torchdata.datapipes.map import SequenceWrapper
+        >>> source_dp = SequenceWrapper(range(10))
+        >>> cache_dp = source_dp.in_memory_cache()
+        >>> list(cache_dp)
+        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+    """
+
+    def __init__(self, source_dp: MapDataPipe[T_co], cache: Optional[Callable] = None) -> None:
```

Comment:
Why do we want to expose `cache` to users? I believe a `dict` should be sufficient for the cache.

Reply:
If it's LRU cache, the element will be popped out the cache when it's full. Then, there is a bug that `index_cached` will still keep the index that has already removed from the cache.

I believe you want to use `numpy.array` or `mp.Array` for COW issue during multiprocessing. However, I don't think this is the problem in this case because `self.cache` is empty when we pass this DataPipe instance into multiprocessing. It's different from putting list of indices into `Dataset` causing the COW issue. This is legitimate to have increasing memory usage as each process will hold a separate `cache` object.

Reply:
If the COW issue isn't relevant here, then I will just use a `dict`. Thanks for the comment!